add h5py to setup.py's extra_requires

orbeckst opened the pull request:
Fixes #3701

Changes made in this Pull Request:
- keep minimum version of h5py at >= 2.1.0
- new 'extra_formats' target in extra_requires where we can add pip-installable packages for additional formats that are not installed by default
- deprecated 'AMBER' target in extra_requires (now part of 'extra_formats')


PR Checklist
------------
 - [ ] Tests? (not sure how to test)
 - [ ] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

orbeckst committed:
add h5py to setup.py's extra_requires

- fix #3701
- new 'extra_coordinates' target in extra_requires where we can
  add pip-installable packages for additional formats that are not
  installed by default
- keep minimum version of h5py at >= 2.1.0

orbeckst commented on package/setup.py:
```diff
@@ -63,15 +63,8 @@
     print('Please upgrade your version of Python.')
     sys.exit(-1)
 
-if sys.version_info[0] < 3:
-    import ConfigParser as configparser
-else:
-    import configparser
-
-if sys.version_info[0] >= 3:
-    from subprocess import getoutput
-else:
-    from commands import getoutput
```
removed Python 2 compatibility crud

orbeckst commented on package/setup.py:
```diff
@@ -650,6 +643,10 @@ def long_description(readme):
               'AMBER': [
                   'netCDF4>=1.0',  # for fast AMBER writing, also needs HDF5
               ],
+              'extra_formats': [   # additional file formats
+                  'netCDF4>=1.0',  # for fast AMBER writing, also needs HDF5
+                  'h5py>=2.10',    # H5MD
+                  ],
```
possible add here
* gsd (unless we just install it anyway)
* chemfiles

IAlibay replied:
gsd is an install requires for now so it shouldn't also be going into extra requires

chemfiles might be an option if we want

orbeckst replied:
I added chemfiles>=0.10

orbeckst commented on package/setup.py:
```diff
@@ -650,6 +643,10 @@ def long_description(readme):
               'AMBER': [
```
deprecated the AMBER target (not sure if anyone ever used it...)

orbeckst commented on package/setup.py:
```diff
@@ -650,6 +643,10 @@ def long_description(readme):
               ],
+              'extra_formats': [   # additional file formats
+                  'netCDF4>=1.0',  # for fast AMBER writing, also needs HDF5
+                  'h5py>=2.10',    # H5MD
```
left at minimal version for right now

IAlibay replied:
This needs to be reflected in all the CI places:
- [x] gh action
- [x] azure
- <s>[ ] Travis</s>
- [x] maintainer's yaml

orbeckst replied:
- Travis does not install h5py. https://github.com/MDAnalysis/mdanalysis/blob/0b3ba59d7a31d461593dfa83ef50b38f171ad9eb/.travis.yml#L27-L28
- Azure does not pin or restrict packages when h5py is installed https://github.com/MDAnalysis/mdanalysis/blob/0b3ba59d7a31d461593dfa83ef50b38f171ad9eb/azure-pipelines.yml#L84-L98 although farther down, we have packages with restrictions (such as chemfiles). I am not sure what to add where... please review

IAlibay replied:
I think it's fine, I'll give it a review once I'm back from the wilds of Edinburgh

orbeckst replied:
I had change the azure file to have the restriction and the tests passed, so at least it didn't immediately break anything.

IAlibay reviewed (changes requested):
oops comments weren't via review - I'll request changes over setting minimum versions for all the CI files please.

orbeckst committed:
minimum h5py >= 2.10

- GitHub actions
- maintainer/conda/environment.yml for readthedocs
- Azure pipelines

orbeckst committed:
update extra_deps in setup.py

- include chemfiles>=0.10 in extra_formats
- added comment to remove AMBER in 3.0.0

orbeckst committed:
updated CHANGELOG

pep8speaks commented:
Hello @orbeckst! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



There are currently no PEP 8 issues detected in this Pull Request. Cheers! :beers: 

##### Comment last updated at 2022-07-05 19:28:49 UTC

orbeckst committed:
pep8 fixes

orbeckst committed:
updated sklearn --> scikit-learn in setup.py

orbeckst commented:
I installed into an empty py39 env with
```
pip install ./package[extra_formats,analysis]
```
and packages import
```
Python 3.9.0 | packaged by conda-forge | (default, Nov 26 2020, 07:54:06)
[Clang 11.0.0 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
imp>>> import h5py
>>> import chemfiles
>>> h5py.__version__
'3.7.0'
>>> chemfiles.__version__
'0.10.2'
>>> import netCDF4
>>> netCDF4.__version__
'1.5.8'
>>> import MDAnalysis
>>> MDAnalysis.__version__
'2.3.0-dev0'
```
and after installing the testsuite,  all tests pass
```
pytest  --disable-warnings  -n 4  testsuite/MDAnalysisTests
```
```
================= 17782 passed, 359 skipped, 2 xfailed, 2 xpassed, 32302 warnings in 315.77s (0:05:15) ==================
```

IAlibay commented on azure-pipelines.yml:
```diff
@@ -92,7 +92,7 @@ jobs:
       scikit-learn
       scipy
-      h5py
+      h5py>=2.10
```
Let's leave it as this for now, long term we might want to come back and put minimum pins on stuff here too (although to be honest we'll just pick up the maximum version here anyways, I guess I'm just being too pedantic...)

orbeckst replied:
I understand your comment as "h5py>=2.10 is fine for now" so I am not changing it.

IAlibay commented on package/CHANGELOG:
```diff
@@ -25,10 +25,15 @@ Enhancements
+    `pip install ./package[extra_formats]` (Issue #3701)
 
 Deprecations
   * Deprecate `density` parameter in favor of `norm` in InterRDF_s
     (Issue #3687)
+  * The extras_requires target "AMBER" for `pip install ./package[AMBER]`
+    will be removed in 3.0. Use "extra_formats". (Issue #3701, PR #3711)
```
```suggestion
  * The extras_requires target "AMBER" for `pip install ./package[AMBER]`
    will be removed in 3.0. Use "extra_formats". (Issue #3701, PR #3711)
  * Deprecate `density` parameter in favor of `norm` in InterRDF_s
    (Issue #3687)
```

Newest entry first :)

IAlibay replied:
Also I guess we shouldn't be a pain to the couple of folks that might use this, but does it really need to be a major version breaking change? The number of automated workflow using this should be near 0, and there's no way they'll get a deprecatewarning about this. If we enact the change the biggest issue is that things will just break cleanly right? (or will it just ignore the keyword and go ahead)

orbeckst replied:
I don’t think it really needs to be a major revision deprecation. I was just very conservative. I’m not sure how many people used it, tbh.

i haven’t tried proving missing/incorrect target names. I *assume* it will break. 

I’m happy to go with a deprecation for 2.4, too, just let me know or change directly in setup.py and CHANGELOG.

orbeckst replied:
I set the deprecation for 2.4.

IAlibay replied:
I tried passing the wrong extra (as 'analyses' instead of 'analysis')
```
WARNING: mdanalysis 2.3.0.dev0 does not provide the extra 'analyses'
````

So it won't fail if you give it the wrong thing (which is rather annoying). I'm still ok with setting the deprecation for 2.4 given the fact that we can't suitably warn users anyways, it'll be a problem either way.

IAlibay reviewed (changes requested):
just the one thing, otherwise lgtm, thanks @orbeckst !

IAlibay commented:
Whilst we're here, do the changes here indicate that this is also fixed? https://github.com/MDAnalysis/mdanalysis/issues/3406

orbeckst committed:
Update package/CHANGELOG

Co-authored-by: Irfan Alibay <IAlibay@users.noreply.github.com>

orbeckst commented:
Regarding #3406 I don’t know — we could add a restriction to not use h5py 3.4.0 but we would be guessing as we never pinpointed the ultimate cause. Ultimately we need someone to test if that failure is really only occurring with specific releases, then we exclude all bad ones here, and then I’d be comfortable closing the issue.

orbeckst committed:
deprecate extras_require "AMBER" target for 2.4

IAlibay committed:
Merge branch 'develop' into h5py-extradep

orbeckst commented:
coverage-related error makes one test fail — no idea what to do about it; is this related to the need to pin coverage again that was mentioned somewhere on discord?

orbeckst commented:
@IAlibay let me know if you think something else needs to be done here. No rush.

orbeckst committed:
Merge branch 'develop' into h5py-extradep

orbeckst commented:
GitHub hangs on "Checking for ability to merge automatically…" —weird. 

Anyway, @IAlibay is there anything else that needs doing?

IAlibay reviewed (approved):
"I'll give it a review once I'm back from the wilds of Edinburgh" oof I think I see why I didn't get around to re-reviewing this until this week. So sorry D: !

IAlibay committed:
Merge branch 'develop' into h5py-extradep
